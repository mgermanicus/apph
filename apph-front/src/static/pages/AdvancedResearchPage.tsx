import * as React from 'react';
import { useEffect, useState } from 'react';
import { FilterSelector } from '../components/FilterSelector';
import { IFilter, IFilterPayload } from '../../utils/types/Filter';
import { AlertSnackbar } from '../components/AlertSnackbar';
import { usePhotoTable } from '../../utils/hooks/usePhotoTable';
import { deepCopy } from '../../utils/DeepCopy';
import { useLocation } from 'react-router-dom';
import TagService from '../../services/TagService';
import { ITag } from '../../utils';
import { setTagList } from '../../redux/slices/tagSlice';
import { useDispatch } from 'react-redux';

export const AdvancedResearchPage = (): JSX.Element => {
  const location = useLocation();
  const dispatch = useDispatch();
  const [filterList, setFilterList] = useState<IFilterPayload[]>([]);
  const { errorState, photoTable } = usePhotoTable(filterList);

  useEffect(() => {
    TagService.getAllTags(
      (tags: string) => {
        const tagsConverted: ITag[] = JSON.parse(tags);
        dispatch(setTagList(tagsConverted));
      },
      (errorMessage: string) => errorState.setMessage(errorMessage)
    );
  }, []);

  const getFilterPayload = (filterList: IFilter[]): void => {
    if (
      filterList.length === 0 ||
      (filterList.some((filterState) =>
        Object.values(filterState).some(
          (value) =>
            value === undefined ||
            value === '' ||
            (Array.isArray(value) && value.length === 0)
        )
      ) &&
        location.state === null)
    ) {
      errorState.setMessage('filter.error.fieldNull');
    } else {
      const filterPayload = filterList as IFilterPayload[];
      setFilterList(deserializeArray(filterPayload));
    }
  };

  const deserializeArray = (
    filterPayload: IFilterPayload[]
  ): IFilterPayload[] => {
    filterPayload.forEach((filter) => {
      delete filter.id;
      if (Array.isArray(filter.value)) {
        const oldTagsArray = filter.value;
        filter.value = filter.value[0].name;
        for (let i = 1; i < oldTagsArray.length; i++) {
          const newFilter = {
            field: filter.field,
            operator: null,
            value: oldTagsArray[i].name
          };
          filterPayload.push(newFilter);
        }
      }
    });
    return filterPayload;
  };

  const handleFilterPhoto = (filterList: IFilter[]) => {
    const filters: IFilter[] = deepCopy(filterList);
    getFilterPayload(filters);
  };

  return (
    <>
      <FilterSelector
        tagName={(location?.state as { tagName: string })?.tagName}
        onFilterPhoto={handleFilterPhoto}
        onError={errorState.setMessage}
      />
      {!!filterList?.length && photoTable}
      <AlertSnackbar
        open={!!errorState.getMessage}
        severity={'warning'}
        message={errorState.getMessage}
        onClose={() => errorState.setMessage('')}
      />
    </>
  );
};
