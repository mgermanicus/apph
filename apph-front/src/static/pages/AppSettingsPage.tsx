import SettingService from '../../services/SettingService';
import { FormEvent, useEffect, useState } from 'react';
import { cardStyle, ISetting } from '../../utils';
import {
  Alert,
  Box,
  Button,
  Card,
  CardContent,
  CardHeader,
  InputAdornment,
  Stack,
  TextField
} from '@mui/material';
import { ConfirmationDialog } from '../components/ConfirmationDialog';
import { useTranslation } from 'react-i18next';

export const AppSettingsPage = () => {
  const { t } = useTranslation();
  const [uploadSize, setUploadSize] = useState<number>(0);
  const [downloadSize, setDownloadSize] = useState<number>(0);
  const [message, setMessage] = useState<string>('');
  const [errorOccured, setErrorOccured] = useState(false);
  const [dialogOpen, setDialogOpen] = useState(false);

  useEffect(() => {
    (async () => {
      await SettingService.getSettings(
        (data) => {
          const settings: ISetting = JSON.parse(data);
          setUploadSize(settings.uploadSize);
          setDownloadSize(settings.downloadSize);
        },
        () => {
          setErrorOccured(true);
          setMessage('admin.setting.error.errorGet');
        }
      );
    })();
  }, []);

  const handleSubmit = async (event: FormEvent) => {
    event.preventDefault();
    if (uploadSize > 0 && downloadSize > 0) {
      setDialogOpen(true);
    } else {
      setMessage('admin.setting.error.valueNotValid');
    }
  };

  const updateParameters = async () =>
    SettingService.updateSettings(
      uploadSize,
      downloadSize,
      () => setMessage('admin.setting.success.modifySuccess'),
      () => {
        setErrorOccured(true);
        setMessage('admin.setting.error.modifyFailed');
      }
    );

  return (
    <Card sx={cardStyle}>
      <CardHeader title={t('admin.setting.edit')} />
      <CardContent>
        <Box component="form" onSubmit={handleSubmit}>
          <Stack spacing={2}>
            <TextField
              required
              label={t('admin.setting.uploadSize')}
              value={uploadSize}
              InputProps={{
                endAdornment: <InputAdornment position="end">MB</InputAdornment>
              }}
              onChange={(e) =>
                setUploadSize(
                  isNaN(parseInt(e.target.value)) ? 0 : parseInt(e.target.value)
                )
              }
            />
            <TextField
              required
              label={t('admin.setting.downloadSize')}
              value={downloadSize}
              InputProps={{
                endAdornment: <InputAdornment position="end">MB</InputAdornment>
              }}
              onChange={(e) =>
                setDownloadSize(
                  isNaN(parseInt(e.target.value)) ? 0 : parseInt(e.target.value)
                )
              }
            />
            <Button type="submit">{t('action.confirm')}</Button>

            {!!message && (
              <Alert severity={errorOccured ? 'error' : 'success'} id="alert">
                {t(message)}
              </Alert>
            )}
          </Stack>
        </Box>
        <ConfirmationDialog
          open={dialogOpen}
          onConfirm={async () => {
            setDialogOpen(false);
            await updateParameters();
          }}
          onCancel={() => {
            setDialogOpen(false);
          }}
          message={t('admin.setting.warningUpdate')}
        />
      </CardContent>
    </Card>
  );
};
