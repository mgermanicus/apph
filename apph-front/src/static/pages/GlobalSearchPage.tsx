import * as React from 'react';
import { useEffect, useState } from 'react';
import { useLocation } from 'react-router-dom';
import PhotoService from '../../services/PhotoService';
import { IMessage, ITable } from '../../utils';
import {
  AlertColor,
  Backdrop,
  Box,
  Button,
  ButtonGroup,
  Grid,
  Pagination,
  Stack,
  Typography
} from '@mui/material';
import { PhotoComplexGrid } from '../components/PhotoComplexGrid';
import CircularProgress from '@mui/material/CircularProgress';
import { useTranslation } from 'react-i18next';
import { AlertSnackbar } from '../components/AlertSnackbar';

export const GlobalSearchPage = ({
  pageSize = 5
}: {
  pageSize?: number;
}): JSX.Element => {
  const { t } = useTranslation();
  const location = useLocation();
  const [data, setData] = useState<ITable[]>([]);
  const [total, setTotal] = React.useState<number>(0);
  const [page, setPage] = React.useState<number>(1);
  const [loading, setLoading] = React.useState<boolean>(false);
  const [message, setMessage] = useState('');
  const [snackbarOpen, setSnackbarOpen] = useState<boolean>(false);
  const [severity, setSeverity] = useState<AlertColor>();
  const [tagFacet, setTagFacet] = useState<Record<string, number>>({});
  const [sizeFacet, setSizeFacet] = useState<Record<any, number>>({});

  const traitError = (error: IMessage) => {
    setMessage(error.message);
    setSnackbarOpen(true);
    setSeverity('error');
  };

  useEffect(() => {
    setLoading(true);
    PhotoService.search(
      location?.search.replace('%2', '/'),
      page,
      pageSize,
      (photoList, totalHits, facets) => {
        setData(photoList);
        setTotal(totalHits);
        setTagFacet(facets.tagFacet);
        setSizeFacet(facets.sizeFacet);
      },
      (error: IMessage) => traitError(error)
    ).finally(() => {
      setLoading(false);
    });
  }, [location, page]);

  const handleChange = (event: React.ChangeEvent<unknown>, value: number) => {
    setPage(value);
    window.scrollTo(0, 0);
  };

  if (total == 0) {
    return (
      <Grid
        container
        direction="row"
        justifyContent="center"
        alignItems="center"
        sx={{ height: '90vh' }}
      >
        <Typography variant="h6" gutterBottom component="div">
          {t('photo.error.notFound')}
        </Typography>
        <AlertSnackbar
          open={snackbarOpen}
          severity={severity}
          message={t(message)}
          onClose={setSnackbarOpen}
        />
      </Grid>
    );
  }
  return (
    <Box>
      <Grid container spacing={1} justifyContent="flex-start">
        <Grid
          item
          sx={{
            width: 'calc(.25 * (100vw - 28px))',
            minWidth: '249px',
            maxWidth: '374px',
            float: 'left'
          }}
        >
          <Typography variant="h5" gutterBottom component="div">
            {t('photo.enhanceSearch')}
          </Typography>
          <Typography variant="h6" gutterBottom component="div" align="left">
            {t('photo.tagSuggestions')}
          </Typography>
          <ButtonGroup
            variant="text"
            aria-label="text button group"
            orientation="vertical"
            sx={{ alignItems: 'baseline', width: 1, mb: '8px' }}
          >
            {Object.entries(tagFacet).map(([key, value]) => (
              <Button
                key={key}
                sx={{
                  justifyContent: 'left',
                  textAlign: 'left !important',
                  maxWidth: 'calc(.25 * (100vw - 28px))',
                  overflow: 'hidden'
                }}
                onClick={() => console.log(key)}
              >
                {key} ({value})
              </Button>
            ))}
          </ButtonGroup>
          <Typography variant="h6" gutterBottom component="div" align="left">
            {t('photo.sizeSuggestions')}
          </Typography>
          <ButtonGroup
            variant="text"
            aria-label="text button group"
            orientation="vertical"
            sx={{ alignItems: 'baseline', width: 1, mb: '8px' }}
          >
            {Object.entries(sizeFacet).map(([key, value], index) => {
              const ranges: Array<string> = key.slice(1, -1).split(',');
              const keyString =
                index !== 2
                  ? `${parseInt(ranges[0])} ko - ${parseInt(ranges[1])} ko
              (${value})`
                  : `> ${parseInt(ranges[0])} ko (${value})`;
              return (
                <Button
                  key={key}
                  sx={{
                    justifyContent: 'left',
                    textAlign: 'left',
                    maxWidth: 'calc(.25 * (100vw - 28px))',
                    overflow: 'hidden'
                  }}
                  onClick={() => console.log(ranges)}
                >
                  {keyString}
                </Button>
              );
            })}
          </ButtonGroup>
        </Grid>
        <Grid item xs>
          <Stack spacing={2}>
            {data.map((photo) => (
              <React.Fragment key={photo.id}>
                <PhotoComplexGrid photo={photo} />
              </React.Fragment>
            ))}
          </Stack>
          <Stack spacing={2}>
            <Pagination
              count={Math.ceil(total / pageSize)}
              showFirstButton
              showLastButton
              size="large"
              sx={{ margin: 'auto', paddingTop: 1, paddingBottom: 1 }}
              page={page}
              onChange={handleChange}
            />
          </Stack>
          <Backdrop
            sx={{ color: '#fff', zIndex: (theme) => theme.zIndex.drawer + 1 }}
            open={loading}
          >
            <CircularProgress color="inherit" />
          </Backdrop>
        </Grid>
      </Grid>
    </Box>
  );
};
