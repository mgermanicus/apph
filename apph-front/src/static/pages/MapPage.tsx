import { Box } from '@mui/material';
import { PhotosMap } from '../components/PhotosMap';

export const MapPage = () => {
  return (
    <Box>
      <PhotosMap locations={[]} />
    </Box>
  );
};
