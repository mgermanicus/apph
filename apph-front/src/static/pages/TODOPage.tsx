import { makeCardStyles } from '../../utils';
import { Card, CardContent, CardHeader, Typography } from '@mui/material';
import * as React from 'react';

//TODO replace this component by the corresponding component page in privateRoutes.tsx
//TODO remove this file when all pages are done
export const TODOPage = ({ todo }: { todo: string }): JSX.Element => {
  const classes = makeCardStyles();
  return (
    <Card className={classes.cardStyle}>
      <CardHeader title="TODO" />
      <CardContent>
        <Typography>{todo}</Typography>
      </CardContent>
    </Card>
  );
};
