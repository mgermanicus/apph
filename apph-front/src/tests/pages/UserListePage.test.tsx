import { render } from '@testing-library/react';
import { UserListPage } from '../../static/pages/UserListPage';

jest.mock('react-i18next', () => ({
  // this mock makes sure any components using the translate hook can use it without a warning being shown
  useTranslation: () => {
    return {
      t: (str: string) => str
    };
  }
}));

describe('UserListPage', () => {
  it('should render', () => {
    //WHEN
    const tree = render(<UserListPage />);
    //THEN
    expect(tree).toMatchSnapshot();
  });
});
