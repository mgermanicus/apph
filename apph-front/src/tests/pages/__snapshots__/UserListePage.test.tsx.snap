// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserListPage should render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="height: 90vh; width: 100%;"
      >
        <div
          aria-colcount="4"
          aria-multiselectable="false"
          aria-rowcount="0"
          class="MuiDataGrid-root MuiDataGrid-root--densityStandard css-r11z79-MuiDataGrid-root"
          role="grid"
        >
          <div>
            <div />
          </div>
          <div
            class="MuiDataGrid-main css-204u17-MuiDataGrid-main"
          >
            <div
              style="height: -56px; width: 0px; position: absolute; top: 56px;"
            >
              <div
                class="MuiDataGrid-overlay css-1lk5ll6-MuiDataGrid-overlay"
              >
                <span
                  class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
                  role="progressbar"
                  style="width: 40px; height: 40px;"
                >
                  <svg
                    class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
                    viewBox="22 22 44 44"
                  >
                    <circle
                      class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                      cx="44"
                      cy="44"
                      fill="none"
                      r="20.2"
                      stroke-width="3.6"
                    />
                  </svg>
                </span>
              </div>
            </div>
            <div
              class="MuiDataGrid-columnHeaders css-f3jnds-MuiDataGrid-columnHeaders"
              style="min-height: 56px; max-height: 56px; line-height: 56px;"
            >
              <div
                aria-rowindex="1"
                class="MuiDataGrid-columnHeadersInner css-1s0hp0k-MuiDataGrid-columnHeadersInner"
                role="row"
                style="transform: translate3d(0px, 0px, 0px);"
              >
                <div
                  aria-colindex="1"
                  aria-sort="none"
                  class="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--alignCenter"
                  data-field="firstname"
                  role="columnheader"
                  style="width: 0px; min-width: 0; max-width: 0;"
                  tabindex="0"
                >
                  <div
                    class="MuiDataGrid-columnHeaderDraggableContainer"
                    draggable="false"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderTitleContainer"
                    >
                      <div
                        class="MuiDataGrid-columnHeaderTitleContainerContent"
                      >
                        <div
                          aria-label=""
                          class="MuiDataGrid-columnHeaderTitle css-1jbbcbn-MuiDataGrid-columnHeaderTitle"
                          data-mui-internal-clone-element="true"
                        >
                          userTable.firstname
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                    style="min-height: 56px; opacity: 1;"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                      data-testid="SeparatorIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M11 19V5h2v14z"
                      />
                    </svg>
                  </div>
                </div>
                <div
                  aria-colindex="2"
                  aria-sort="none"
                  class="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--alignCenter"
                  data-field="lastname"
                  role="columnheader"
                  style="width: 0px; min-width: 0; max-width: 0;"
                  tabindex="-1"
                >
                  <div
                    class="MuiDataGrid-columnHeaderDraggableContainer"
                    draggable="false"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderTitleContainer"
                    >
                      <div
                        class="MuiDataGrid-columnHeaderTitleContainerContent"
                      >
                        <div
                          aria-label=""
                          class="MuiDataGrid-columnHeaderTitle css-1jbbcbn-MuiDataGrid-columnHeaderTitle"
                          data-mui-internal-clone-element="true"
                        >
                          userTable.lastname
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                    style="min-height: 56px; opacity: 1;"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                      data-testid="SeparatorIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M11 19V5h2v14z"
                      />
                    </svg>
                  </div>
                </div>
                <div
                  aria-colindex="3"
                  aria-sort="none"
                  class="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--alignCenter"
                  data-field="email"
                  role="columnheader"
                  style="width: 0px; min-width: 0; max-width: 0;"
                  tabindex="-1"
                >
                  <div
                    class="MuiDataGrid-columnHeaderDraggableContainer"
                    draggable="false"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderTitleContainer"
                    >
                      <div
                        class="MuiDataGrid-columnHeaderTitleContainerContent"
                      >
                        <div
                          aria-label=""
                          class="MuiDataGrid-columnHeaderTitle css-1jbbcbn-MuiDataGrid-columnHeaderTitle"
                          data-mui-internal-clone-element="true"
                        >
                          userTable.email
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                    style="min-height: 56px; opacity: 1;"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                      data-testid="SeparatorIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M11 19V5h2v14z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
            <div
              style="overflow: visible; height: 0px; width: 0px;"
            >
              <div
                class="MuiDataGrid-virtualScroller css-1w5m2wr-MuiDataGrid-virtualScroller"
                style="width: 0px; height: auto; margin-top: 56px; overflow-x: hidden;"
              >
                <div
                  class="MuiDataGrid-virtualScrollerContent MuiDataGrid-virtualScrollerContent--overflowed css-1kwdphh-MuiDataGrid-virtualScrollerContent"
                  style="width: auto; height: 1px; min-height: auto;"
                >
                  <div
                    class="MuiDataGrid-virtualScrollerRenderZone css-s1v7zr-MuiDataGrid-virtualScrollerRenderZone"
                    style="transform: translate3d(0px, undefinedpx, 0px);"
                  />
                </div>
              </div>
            </div>
            <div
              class="Mui-resizeTriggers"
            >
              <div
                class="expand-trigger"
              >
                <div
                  style="width: 1px; height: 1px;"
                />
              </div>
              <div
                class="contract-trigger"
              />
            </div>
          </div>
          <div>
            <div
              class="MuiDataGrid-footerContainer css-17jjc08-MuiDataGrid-footerContainer"
            >
              <div />
              <div
                class="MuiTablePagination-root css-rtrcn9-MuiTablePagination-root"
              >
                <div
                  class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-11bfvty-MuiToolbar-root-MuiTablePagination-toolbar"
                >
                  <div
                    class="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
                  />
                  <p
                    class="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
                  >
                    0â€“0 of 0
                  </p>
                  <div
                    class="MuiTablePagination-actions"
                  >
                    <button
                      aria-label="Go to previous page"
                      class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                      disabled=""
                      tabindex="-1"
                      title="Go to previous page"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="KeyboardArrowLeftIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
                        />
                      </svg>
                    </button>
                    <button
                      aria-label="Go to next page"
                      class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                      disabled=""
                      tabindex="-1"
                      title="Go to next page"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="KeyboardArrowRightIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
                        />
                      </svg>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="height: 90vh; width: 100%;"
    >
      <div
        aria-colcount="4"
        aria-multiselectable="false"
        aria-rowcount="0"
        class="MuiDataGrid-root MuiDataGrid-root--densityStandard css-r11z79-MuiDataGrid-root"
        role="grid"
      >
        <div>
          <div />
        </div>
        <div
          class="MuiDataGrid-main css-204u17-MuiDataGrid-main"
        >
          <div
            style="height: -56px; width: 0px; position: absolute; top: 56px;"
          >
            <div
              class="MuiDataGrid-overlay css-1lk5ll6-MuiDataGrid-overlay"
            >
              <span
                class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
                role="progressbar"
                style="width: 40px; height: 40px;"
              >
                <svg
                  class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
                  viewBox="22 22 44 44"
                >
                  <circle
                    class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                    cx="44"
                    cy="44"
                    fill="none"
                    r="20.2"
                    stroke-width="3.6"
                  />
                </svg>
              </span>
            </div>
          </div>
          <div
            class="MuiDataGrid-columnHeaders css-f3jnds-MuiDataGrid-columnHeaders"
            style="min-height: 56px; max-height: 56px; line-height: 56px;"
          >
            <div
              aria-rowindex="1"
              class="MuiDataGrid-columnHeadersInner css-1s0hp0k-MuiDataGrid-columnHeadersInner"
              role="row"
              style="transform: translate3d(0px, 0px, 0px);"
            >
              <div
                aria-colindex="1"
                aria-sort="none"
                class="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--alignCenter"
                data-field="firstname"
                role="columnheader"
                style="width: 0px; min-width: 0; max-width: 0;"
                tabindex="0"
              >
                <div
                  class="MuiDataGrid-columnHeaderDraggableContainer"
                  draggable="false"
                >
                  <div
                    class="MuiDataGrid-columnHeaderTitleContainer"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderTitleContainerContent"
                    >
                      <div
                        aria-label=""
                        class="MuiDataGrid-columnHeaderTitle css-1jbbcbn-MuiDataGrid-columnHeaderTitle"
                        data-mui-internal-clone-element="true"
                      >
                        userTable.firstname
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                  style="min-height: 56px; opacity: 1;"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                    data-testid="SeparatorIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M11 19V5h2v14z"
                    />
                  </svg>
                </div>
              </div>
              <div
                aria-colindex="2"
                aria-sort="none"
                class="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--alignCenter"
                data-field="lastname"
                role="columnheader"
                style="width: 0px; min-width: 0; max-width: 0;"
                tabindex="-1"
              >
                <div
                  class="MuiDataGrid-columnHeaderDraggableContainer"
                  draggable="false"
                >
                  <div
                    class="MuiDataGrid-columnHeaderTitleContainer"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderTitleContainerContent"
                    >
                      <div
                        aria-label=""
                        class="MuiDataGrid-columnHeaderTitle css-1jbbcbn-MuiDataGrid-columnHeaderTitle"
                        data-mui-internal-clone-element="true"
                      >
                        userTable.lastname
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                  style="min-height: 56px; opacity: 1;"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                    data-testid="SeparatorIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M11 19V5h2v14z"
                    />
                  </svg>
                </div>
              </div>
              <div
                aria-colindex="3"
                aria-sort="none"
                class="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--alignCenter"
                data-field="email"
                role="columnheader"
                style="width: 0px; min-width: 0; max-width: 0;"
                tabindex="-1"
              >
                <div
                  class="MuiDataGrid-columnHeaderDraggableContainer"
                  draggable="false"
                >
                  <div
                    class="MuiDataGrid-columnHeaderTitleContainer"
                  >
                    <div
                      class="MuiDataGrid-columnHeaderTitleContainerContent"
                    >
                      <div
                        aria-label=""
                        class="MuiDataGrid-columnHeaderTitle css-1jbbcbn-MuiDataGrid-columnHeaderTitle"
                        data-mui-internal-clone-element="true"
                      >
                        userTable.email
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                  style="min-height: 56px; opacity: 1;"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                    data-testid="SeparatorIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M11 19V5h2v14z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
          <div
            style="overflow: visible; height: 0px; width: 0px;"
          >
            <div
              class="MuiDataGrid-virtualScroller css-1w5m2wr-MuiDataGrid-virtualScroller"
              style="width: 0px; height: auto; margin-top: 56px; overflow-x: hidden;"
            >
              <div
                class="MuiDataGrid-virtualScrollerContent MuiDataGrid-virtualScrollerContent--overflowed css-1kwdphh-MuiDataGrid-virtualScrollerContent"
                style="width: auto; height: 1px; min-height: auto;"
              >
                <div
                  class="MuiDataGrid-virtualScrollerRenderZone css-s1v7zr-MuiDataGrid-virtualScrollerRenderZone"
                  style="transform: translate3d(0px, undefinedpx, 0px);"
                />
              </div>
            </div>
          </div>
          <div
            class="Mui-resizeTriggers"
          >
            <div
              class="expand-trigger"
            >
              <div
                style="width: 1px; height: 1px;"
              />
            </div>
            <div
              class="contract-trigger"
            />
          </div>
        </div>
        <div>
          <div
            class="MuiDataGrid-footerContainer css-17jjc08-MuiDataGrid-footerContainer"
          >
            <div />
            <div
              class="MuiTablePagination-root css-rtrcn9-MuiTablePagination-root"
            >
              <div
                class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-11bfvty-MuiToolbar-root-MuiTablePagination-toolbar"
              >
                <div
                  class="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
                />
                <p
                  class="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
                >
                  0â€“0 of 0
                </p>
                <div
                  class="MuiTablePagination-actions"
                >
                  <button
                    aria-label="Go to previous page"
                    class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                    disabled=""
                    tabindex="-1"
                    title="Go to previous page"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="KeyboardArrowLeftIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
                      />
                    </svg>
                  </button>
                  <button
                    aria-label="Go to next page"
                    class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                    disabled=""
                    tabindex="-1"
                    title="Go to next page"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="KeyboardArrowRightIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
                      />
                    </svg>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
